File: PE_0002/PE_0002.m
function [ sumeven ] = PE_0002( maxf )
%   PE_0002

%   Even Fibonacci numbers

%   sum of all even valued members of the Fibonacci sequence that are less than 4 million

    terms=[];
    f=0;
    sumeven=0;
    n=1;
    while f < maxf
        f = fibo(n);
        terms=[terms,f];
        sumeven=sum(terms(mod(terms,2)==0)) ;
        n=n+1;
    end
end
   
function f = fibo(n)
    if ( ~isscalar(n) || n < 1 || n ~= fix(n))
        error('n must be a positive integer!');
    end
    if n <= 2
        f = 1;
    else
        f = fibo(n-2) + fibo(n-1);
    end

end

File: PE_0002/PE_0002.py
# -*- coding: utf-8 -*-
"""
PE_0002

Even Fibonacci numbers

By considering the terms in the Fibonacci sequence whose values do not exceed 
four million, find the sum of the even-valued terms.

Created on Mon Sep 26 15:54:34 2016
@author: mbh
"""
from timeit import default_timer as timer

#Guttag p 254
def fastFib(n, memo = {}):
    """Assumes n is an int >= 0, memo used only by recursive calls
       Returns Fibonacci of n"""
    if n == 0 or n == 1:
        return 1
    try:
        return memo[n]
    except KeyError:
        result = fastFib(n-1, memo) + fastFib(n-2, memo)
        memo[n] = result
        return result


def p2(limit):
    start=timer()
    x=0
    fibsum=0
    while 1:
        nextx=fastFib(x)
        if nextx>limit:
            break
        if nextx%2==0:
            fibsum+=nextx
        x+=1
    print(fibsum)
    print('Elapsed time:',timer()-start)
        
        
